test/releasers/dotnet.ts

```diff
@@ -0,0 +1,148 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {describe, it, afterEach} from 'mocha';
+import * as nock from 'nock';
+import {DotNet} from '../../src/releasers/dotnet';
+import {readFileSync} from 'fs';
+import {resolve} from 'path';
+import * as snapshot from 'snap-shot-it';
+import * as suggester from 'code-suggester';
+import * as sinon from 'sinon';
+
+const sandbox = sinon.createSandbox();
+const fixturesPath = './test/releasers/fixtures/dotnet';
+
+function getBlob(filename: string) {
+  const content = readFileSync(resolve(fixturesPath, filename), 'utf8').replace(
+    /\r\n/g,
+    '\n'
+  );
+  return {
+    content: Buffer.from(content, 'utf8').toString('base64'),
+    sha: 'abc123',
+  };
+}
+
+describe('DotNet', () => {
+  afterEach(() => {
+    sandbox.restore();
+  });
+  describe('run', () => {
+    it('creates a release PR', async () => {
+      // We stub the entire suggester API, asserting only that the
+      // the appropriate changes are proposed:
+      let expectedChanges = null;
+      sandbox.replace(
+        suggester,
+        'createPullRequest',
+        (_octokit, changes): Promise<number> => {
+          expectedChanges = [...(changes as Map<string, object>)]; // Convert map to key/value pairs.
+          return Promise.resolve(22);
+        }
+      );
+      const graphql = JSON.parse(
+        readFileSync(resolve(fixturesPath, 'commits.json'), 'utf8')
+      );
+      const req = nock('https://api.github.com')
```

Comment:
This is intense.  No request to change it for now, but ... is testing like this a good thing?

Reply:
I think it's been working fairly well as a single end to end smoke test for each release strategy.

For application logic like, the logic around version bumping, tests are self contained unit tests. The alternative is using less `private` methods, and doing more mocking using `sinon`.... but I like these end to end smoke tests.